core/src/main/java/org/apache/calcite/sql/validate/WithScope.java

```diff
@@ -70,4 +73,8 @@ class WithScope extends ListScope {
     }
     super.resolveTable(names, nameMatcher, path, resolved);
   }
+
+  public @Nullable WithRecursiveScope getRecursiveScope() {
+    return recursiveScope;
+  }
 }
```

Comment:
Since the class is not `public`, there's no need for a `public` method. I'd remove this method, and just make the `recursiveScope` field package-protected.